File: .claude/skills/agentteams/scripts/mjs/src/dag.mjs
// Dependency DAG construction, Kahn's topo sort, cycle detection, P0/P1/P2 prioritization.

const RISK_HINTS = ['security', 'data loss', 'regression', 'incident', 'compliance'];
const DEP_PHRASES = [
  /depends on\s+([A-Za-z0-9_/.\- #]+)/gi,
  /blocked by\s+([A-Za-z0-9_/.\- #]+)/gi,
  /requires\s+([A-Za-z0-9_/.\- #]+?)\s+to land/gi,
  /after\s+([A-Za-z0-9_/.\- #]+?)\s+lands/gi,
  /consumes\s+([A-Za-z0-9_/.\- #]+)/gi,
];

function nodeId(featureId, storyId) {
  return `F${featureId}.S${storyId}`;
}

function extractInlineDeps(text) {
  if (!text) return [];
  const ids = new Set();
  for (const re of DEP_PHRASES) {
    let m;
    re.lastIndex = 0;
    while ((m = re.exec(text)) !== null) {
      const ref = m[1].trim();
      const ext = ref.match(/F(\d+)[ .\-]?S?(\d+)?/i);
      if (ext) {
        if (ext[2]) ids.add(`F${ext[1].padStart(2, '0')}.S${ext[2].padStart(2, '0')}`);
        else ids.add(`F${ext[1].padStart(2, '0')}`);
      }
    }
  }
  return [...ids];
}

function normalizeDepRef(raw, currentFeatureId) {
  const m = raw.match(/F(\d+)[ .\-/]?S?(\d+)?/i);
  if (m) {
    if (m[2]) return `F${m[1].padStart(2, '0')}.S${m[2].padStart(2, '0')}`;
    return `F${m[1].padStart(2, '0')}`;
  }
  const sm = raw.match(/S(\d+)/i);
  if (sm && currentFeatureId) return `F${currentFeatureId}.S${sm[1].padStart(2, '0')}`;
  return null;
}

export function buildDag(plan, opts = {}) {
  const implicit = opts.implicit !== false;
  const nodes = [];
  const edges = [];
  const nodeMap = new Map();

  for (const feature of plan.features) {
    for (const story of feature.stories) {
      const id = nodeId(feature.id, story.id);
      const node = {
        id,
        type: 'story',
        featureId: feature.id,
        storyId: story.id,
        slug: story.slug,
        path: story.path,
        acceptanceCriteria: story.acceptanceCriteria,
        taskCount: story.tasks.length,
        risk: detectRisk(story),
      };
      nodes.push(node);
      nodeMap.set(id, node);
    }
  }

  for (const feature of plan.features) {
    for (let i = 0; i < feature.stories.length; i++) {
      const story = feature.stories[i];
      const id = nodeId(feature.id, story.id);
      const featurePadded = feature.id.padStart(2, '0');
      const explicit = (story.dependencies || []).map(d => normalizeDepRef(d, featurePadded)).filter(Boolean);
      const inline = extractInlineDeps(story.context).map(d => {
        return d.includes('.') ? d : null;
      }).filter(Boolean);

      const declaredNone = /none/i.test((story.dependencies || []).join(' ').trim()) || (story.dependencies || []).length === 0;

      const allDeps = new Set([...explicit, ...inline]);

      if (implicit && declaredNone && i > 0) {
        const prev = feature.stories[i - 1];
        allDeps.add(nodeId(feature.id, prev.id));
      } else if (implicit && !declaredNone && i > 0) {
        // explicit deps win, but still add prior story if not declared none
      }

      // Feature-level implicit: F02.S01 depends on F01's last story
      if (implicit && story.id === '01' && plan.features.indexOf(feature) > 0) {
        const prevFeature = plan.features[plan.features.indexOf(feature) - 1];
        if (prevFeature.stories.length > 0) {
          const lastPrev = prevFeature.stories[prevFeature.stories.length - 1];
          allDeps.add(nodeId(prevFeature.id, lastPrev.id));
        }
      }

      for (const from of allDeps) {
        if (nodeMap.has(from) && from !== id) {
          edges.push({ from, to: id });
        }
      }
    }
  }

  return { nodes, edges, nodeMap };
}

function detectRisk(story) {
  const hay = (story.context + ' ' + story.acceptanceCriteria.join(' ')).toLowerCase();
  return RISK_HINTS.filter(h => hay.includes(h));
}

export function topoSort(dag) {
  const { nodes, edges } = dag;
  const indeg = new Map();
  const adj = new Map();
  for (const n of nodes) {
    indeg.set(n.id, 0);
    adj.set(n.id, []);
  }
  for (const e of edges) {
    if (!indeg.has(e.to) || !adj.has(e.from)) continue;
    indeg.set(e.to, indeg.get(e.to) + 1);
    adj.get(e.from).push(e.to);
  }
  const queue = [];
  for (const [id, d] of indeg) if (d === 0) queue.push(id);
  queue.sort();
  const order = [];
  while (queue.length) {
    const cur = queue.shift();
    order.push(cur);
    for (const next of adj.get(cur) || []) {
      indeg.set(next, indeg.get(next) - 1);
      if (indeg.get(next) === 0) {
        queue.push(next);
        queue.sort();
      }
    }
  }
  const cycles = order.length === nodes.length ? [] : findCycles(dag);
  return { order, cycles };
}

function findCycles(dag) {
  const { nodes, edges } = dag;
  const adj = new Map();
  for (const n of nodes) adj.set(n.id, []);
  for (const e of edges) (adj.get(e.from) || []).push(e.to);
  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Map(nodes.map(n => [n.id, WHITE]));
  const cycles = [];
  const stack = [];

  function dfs(u) {
    color.set(u, GRAY);
    stack.push(u);
    for (const v of adj.get(u) || []) {
      if (color.get(v) === GRAY) {
        const idx = stack.indexOf(v);
        cycles.push(stack.slice(idx).concat([v]));
      } else if (color.get(v) === WHITE) {
        dfs(v);
      }
    }
    stack.pop();
    color.set(u, BLACK);
  }
  for (const n of nodes) if (color.get(n.id) === WHITE) dfs(n.id);
  return cycles;
}

export function prioritize(dag, topo) {
  const { nodes, edges } = dag;
  // count downstream blockers (transitive)
  const adj = new Map();
  for (const n of nodes) adj.set(n.id, []);
  for (const e of edges) (adj.get(e.from) || []).push(e.to);
  const downstream = new Map();
  function count(id, seen = new Set()) {
    if (downstream.has(id)) return downstream.get(id);
    if (seen.has(id)) return 0;
    seen.add(id);
    let c = 0;
    for (const v of adj.get(id) || []) {
      c += 1 + count(v, seen);
    }
    downstream.set(id, c);
    return c;
  }
  for (const n of nodes) count(n.id);

  const out = nodes.map(n => {
    const blockers = downstream.get(n.id) || 0;
    let tier = 'P2';
    if (blockers >= 3) tier = 'P1';
    if (blockers >= 6) tier = 'P0';
    if (n.risk.length > 0 && tier === 'P1') tier = 'P0';
    return { ...n, blockersDownstream: blockers, tier };
  });
  out.sort((a, b) => {
    const order = { P0: 0, P1: 1, P2: 2 };
    if (order[a.tier] !== order[b.tier]) return order[a.tier] - order[b.tier];
    if (b.blockersDownstream !== a.blockersDownstream) return b.blockersDownstream - a.blockersDownstream;
    return a.id.localeCompare(b.id);
  });
  return out;
}

File: .claude/skills/agentteams/scripts/py/src/agentteams/dag.py
"""Dependency DAG construction, Kahn's topo sort, cycle detection, P0/P1/P2 prioritization."""
from __future__ import annotations

import re
from typing import Any

RISK_HINTS = ["security", "data loss", "regression", "incident", "compliance"]
DEP_PHRASES = [
    re.compile(r"depends on\s+([A-Za-z0-9_/.\- #]+)", re.IGNORECASE),
    re.compile(r"blocked by\s+([A-Za-z0-9_/.\- #]+)", re.IGNORECASE),
    re.compile(r"requires\s+([A-Za-z0-9_/.\- #]+?)\s+to land", re.IGNORECASE),
    re.compile(r"after\s+([A-Za-z0-9_/.\- #]+?)\s+lands", re.IGNORECASE),
    re.compile(r"consumes\s+([A-Za-z0-9_/.\- #]+)", re.IGNORECASE),
]


def _node_id(feature_id: str, story_id: str) -> str:
    return f"F{feature_id}.S{story_id}"


def _extract_inline_deps(text: str) -> list[str]:
    if not text:
        return []
    ids: set[str] = set()
    for pat in DEP_PHRASES:
        for m in pat.finditer(text):
            ref = m.group(1).strip()
            ext = re.match(r"F(\d+)[ .\-]?S?(\d+)?", ref, re.IGNORECASE)
            if ext:
                if ext.group(2):
                    ids.add(f"F{ext.group(1).zfill(2)}.S{ext.group(2).zfill(2)}")
                else:
                    ids.add(f"F{ext.group(1).zfill(2)}")
    return list(ids)


def _normalize_dep_ref(raw: str, current_feature_id: str | None) -> str | None:
    m = re.search(r"F(\d+)[ .\-/]?S?(\d+)?", raw, re.IGNORECASE)
    if m:
        if m.group(2):
            return f"F{m.group(1).zfill(2)}.S{m.group(2).zfill(2)}"
        return f"F{m.group(1).zfill(2)}"
    sm = re.search(r"S(\d+)", raw, re.IGNORECASE)
    if sm and current_feature_id:
        return f"F{current_feature_id}.S{sm.group(1).zfill(2)}"
    return None


def build_dag(plan: dict[str, Any], implicit: bool = True) -> dict[str, Any]:
    nodes: list[dict[str, Any]] = []
    edges: list[dict[str, str]] = []
    node_map: dict[str, dict[str, Any]] = {}

    for feature in plan["features"]:
        for story in feature["stories"]:
            nid = _node_id(feature["id"], story["id"])
            node = {
                "id": nid,
                "type": "story",
                "featureId": feature["id"],
                "storyId": story["id"],
                "slug": story["slug"],
                "path": story["path"],
                "acceptanceCriteria": story["acceptanceCriteria"],
                "taskCount": len(story["tasks"]),
                "risk": _detect_risk(story),
            }
            nodes.append(node)
            node_map[nid] = node

    for fi, feature in enumerate(plan["features"]):
        for i, story in enumerate(feature["stories"]):
            nid = _node_id(feature["id"], story["id"])
            feature_padded = feature["id"].zfill(2)
            explicit_raw = [_normalize_dep_ref(d, feature_padded) for d in (story.get("dependencies") or [])]
            explicit = [d for d in explicit_raw if d is not None]
            inline = [d for d in _extract_inline_deps(story.get("context", "")) if "." in d]
            declared_none = not (story.get("dependencies") or [])

            all_deps: set[str] = set(explicit) | set(inline)

            if implicit and declared_none and i > 0:
                prev = feature["stories"][i - 1]
                all_deps.add(_node_id(feature["id"], prev["id"]))

            if implicit and story["id"] == "01" and fi > 0:
                prev_feature = plan["features"][fi - 1]
                if prev_feature["stories"]:
                    last_prev = prev_feature["stories"][-1]
                    all_deps.add(_node_id(prev_feature["id"], last_prev["id"]))

            for src in all_deps:
                if src in node_map and src != nid:
                    edges.append({"from": src, "to": nid})

    return {"nodes": nodes, "edges": edges, "nodeMap": node_map}


def _detect_risk(story: dict[str, Any]) -> list[str]:
    hay = (story.get("context", "") + " " + " ".join(story["acceptanceCriteria"])).lower()
    return [h for h in RISK_HINTS if h in hay]


def topo_sort(dag: dict[str, Any]) -> dict[str, Any]:
    nodes = dag["nodes"]
    edges = dag["edges"]
    indeg: dict[str, int] = {n["id"]: 0 for n in nodes}
    adj: dict[str, list[str]] = {n["id"]: [] for n in nodes}
    for e in edges:
        if e["to"] not in indeg or e["from"] not in adj:
            continue
        indeg[e["to"]] += 1
        adj[e["from"]].append(e["to"])
    queue = sorted([i for i, d in indeg.items() if d == 0])
    order: list[str] = []
    while queue:
        cur = queue.pop(0)
        order.append(cur)
        for nxt in adj.get(cur, []):
            indeg[nxt] -= 1
            if indeg[nxt] == 0:
                queue.append(nxt)
                queue.sort()
    cycles = [] if len(order) == len(nodes) else _find_cycles(dag)
    return {"order": order, "cycles": cycles}


def _find_cycles(dag: dict[str, Any]) -> list[list[str]]:
    nodes = dag["nodes"]
    edges = dag["edges"]
    adj: dict[str, list[str]] = {n["id"]: [] for n in nodes}
    for e in edges:
        if e["from"] in adj:
            adj[e["from"]].append(e["to"])
    WHITE, GRAY, BLACK = 0, 1, 2
    color: dict[str, int] = {n["id"]: WHITE for n in nodes}
    cycles: list[list[str]] = []
    stack: list[str] = []

    def dfs(u: str) -> None:
        color[u] = GRAY
        stack.append(u)
        for v in adj.get(u, []):
            if color.get(v) == GRAY:
                idx = stack.index(v)
                cycles.append(stack[idx:] + [v])
            elif color.get(v) == WHITE:
                dfs(v)
        stack.pop()
        color[u] = BLACK

    for n in nodes:
        if color[n["id"]] == WHITE:
            dfs(n["id"])
    return cycles


def prioritize(dag: dict[str, Any], topo: dict[str, Any]) -> list[dict[str, Any]]:
    nodes = dag["nodes"]
    edges = dag["edges"]
    adj: dict[str, list[str]] = {n["id"]: [] for n in nodes}
    for e in edges:
        adj.setdefault(e["from"], []).append(e["to"])
    downstream: dict[str, int] = {}

    def count(nid: str, seen: set[str] | None = None) -> int:
        if nid in downstream:
            return downstream[nid]
        if seen is None:
            seen = set()
        if nid in seen:
            return 0
        seen.add(nid)
        c = 0
        for v in adj.get(nid, []):
            c += 1 + count(v, seen)
        downstream[nid] = c
        return c

    for n in nodes:
        count(n["id"])

    out: list[dict[str, Any]] = []
    for n in nodes:
        blockers = downstream.get(n["id"], 0)
        tier = "P2"
        if blockers >= 3:
            tier = "P1"
        if blockers >= 6:
            tier = "P0"
        if n["risk"] and tier == "P1":
            tier = "P0"
        out.append({**n, "blockersDownstream": blockers, "tier": tier})
    order_map = {"P0": 0, "P1": 1, "P2": 2}
    out.sort(key=lambda r: (order_map[r["tier"]], -r["blockersDownstream"], r["id"]))
    return out
